lib/vector-common/src/estimated_json_encoded_size_of.rs

```diff
@@ -0,0 +1,1039 @@
+            //    requiring allocations before counting the bytes.
+            //
+            // 2. Even for valid UTF-8 encoded bytes, it will have to validate all individual bytes,
+            //    which our soaks have shown to cause a significant drop in throughput.
+            //
+            // Because of this, we take the assumption that all bytes passed through this serializer
+            // are valid UTF-8 encoded bytes, and thus can be counted as-is. If this is not the
+            // case, the final byte size will be off slightly, or significantly, depending on how
+            // many of the bytes need to be escaped, or replaced.
+            Value::Bytes(b) => serializer.serialize_bytes(b),
+
+            // All other `Value` variants are serialized according to the default serialization
+            // implementation of that type.
+            v => v.serialize(serializer),
+        }
+    }
+}
+
+/// A helper trait that is implemented for any `T` that implements `serde::Serialize`, to get the
+/// estimated JSON encoded size of that type.
+pub trait EstimatedJsonEncodedSizeOf {
+    fn estimated_json_encoded_size_of(&self) -> usize;
+}
+
+impl<T> EstimatedJsonEncodedSizeOf for T
+where
+    T: serde::Serialize,
+{
+    /// Returns the estimated JSON encoded size of `self`, or `0` if the size cannot be calculated
+    /// because `T` errors during serialization.
+    #[inline]
+    fn estimated_json_encoded_size_of(&self) -> usize {
+        estimated_size_of(self)
+    }
+}
+
+#[derive(Debug)]
+pub struct Error;
+
+impl std::fmt::Display for Error {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Error")
+    }
+}
+
+impl std::error::Error for Error {}
+impl ser::Error for Error {
+    fn custom<T: std::fmt::Display>(_msg: T) -> Self {
+        Self
+    }
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// A serializer that counts the number of JSON-encoded bytes in a serializable type.
+///
+/// See [`estimated_size_of`] for an easy-to-use wrapper function around this serializer.
+///
+/// This serializer is **optimized for performance**. This means that it is allowed to *approximate*
+/// the size of a type, if doing an exact calculation is too expensive.
+///
+/// Specifically
+///
+/// Additionally, the serializer assumes the type is serialized to a JSON-encoded string using the
+/// `serde_json` crate, which internally uses the `ryu` crate to encode floating point types.
+pub struct Serializer {
+    bytes: usize,
+    start_collection: bool,
+}
+
+/// Return the estimated size of `T` as represented by a JSON-encoded string.
+///
+/// See [`Serializer`] for more details.
+///
+/// # Errors
+///
+/// Returns an error if `T` cannot be serialized.
+pub fn estimated_size_of<T>(value: &T) -> usize
+where
+    T: Serialize,
+{
+    let mut serializer = Serializer {
+        bytes: 0,
+        start_collection: false,
+    };
```

Comment:
Note that if you were to add `#[derive(Default)]` to `Serializer`, this is just `Serializer::default()`.